x-pack/plugins/task_manager/server/saved_objects/index.ts

```diff
@@ -22,5 +24,48 @@ export function setupSavedObjects(
     mappings: mappings.task as SavedObjectsTypeMappingDefinition,
     migrations,
     indexPattern: config.index,
+    deleteOnUpgrade: async (client) => {
+      const oldestNeededActionParams = await getOldestIdleActionTask(client, config.index);
+
+      // Delete all action tasks that have failed and are no longer needed
+      return {
+        bool: {
+          must: [
+            {
+              terms: {
+                'task.taskType': [
+                  'actions:.email',
+                  'actions:.index',
+                  'actions:.pagerduty',
+                  'actions:.swimlane',
+                  'actions:.server-log',
+                  'actions:.slack',
+                  'actions:.webhook',
+                  'actions:.servicenow',
+                  'actions:.servicenow-sir',
+                  'actions:.jira',
+                  'actions:.resilient',
+                  'actions:.teams',
+                ],
+              },
+            },
+            {
+              term: { type: 'task' },
+            },
+            {
+              term: { 'task.status': 'failed' },
+            },
+            {
+              range: {
+                'task.runAt': {
+                  // Only apply to tasks that were run before the oldest needed action
+                  lt: oldestNeededActionParams,
```

Comment:
@mikecote I changed this filter `lt` instead of `lte` because I don't think we want to delete the params that are needed for the oldest idle task? Same for the filter on failed tasks.

Reply:
Yup that makes sense to me 👍 Thanks!